README.md
<!-- readme: start -->
<!-- title: start -->
<br>

<img src = "https://github.com/EnochKang/rankinma/blob/main/vignettes/rankinma_logo.png?raw=true" align = "left" width = "120" />
<br>

# &nbsp; **rankinma** 

#### &nbsp; &nbsp; *Rank in Network Meta-Analysis*

[Chiefeng Chen](https://orcid.org/0000-0002-1595-6553), [Enoch
Kang](https://orcid.org/0000-0002-4903-942X), [Wen-Hsuan
Hou](https://orcid.org/0000-0002-4376-6298), [Jin-Hua
Chen](https://orcid.org/0000-0002-3130-4125), [Yu-Chieh
Chuang](https://orcid.org/0000-0002-7124-6556), & [Edwin
Chan](https://www.duke-nus.edu.sg/core/about/people-leadership/core-visiting-experts/edwin-chan-shih-yen)
<br>
<br>

<!-- title: end -->
<!-- badges: start -->

&nbsp; [![CRAN](https://img.shields.io/cran/v/rankinma?color=blue&label=CRAN&logo=r&logoColor=skyblue)](https://cran.r-project.org/package=rankinma)
&nbsp; [![Update date](https://img.shields.io/badge/Update%20date-2023.09.01-blue.svg?logo=r&logoColor=skyblue)](https://cran.r-project.org/web/packages/rankinma/news/news.html)
&nbsp; [![Shiny](https://img.shields.io/badge/Shiny-shinyapps.io-blue?style=flat&labelColor=gray25&logo=RStudio&logoColor=skyblue)](https://rankinma.shinyapps.io/rankinma/)
&nbsp; [![Licence](https://img.shields.io/badge/licence-GPL--3-brightgreen.svg?color=blue&label=Licence&logo=gnu&logoColor=skyblue)](https://www.gnu.org/licenses/gpl-3.0.en.html)

&nbsp; [![Lifecycle: stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg?color=green&label=Lifecycle&logo=r&logoColor=green)](https://lifecycle.r-lib.org/articles/stages.html#stable)
&nbsp; [![Functions](https://img.shields.io/badge/Functions-7-green.svg?logo=r&logoColor=green)](https://cran.r-project.org/web/packages/rankinma/rankinma.pdf)
&nbsp; [![Dependencis](https://tinyverse.netlify.com/badge/rankinma)](https://cran.r-project.org/package=rankinma)
&nbsp; [![Monthly Downloads](https://cranlogs.r-pkg.org:443/badges/rankinma?color=orange)](https://cranlogs.r-pkg.org:443/badges/rankinma)
&nbsp; [![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/rankinma?color=orange)](https://cran.r-project.org/package=rankinma)

<!-- badges: end -->
<!-- content: start -->
<br>

-   [About rankinma](#about-rankinma)
-   [Feature](#features)
-   [Installation](#installation)
-   [Examples](#examples)
-   [Coding conventions](#coding-conventions)
-   [License](#license)
-   [To do list](#to-do-list)

<!-- content: end -->
<!-- about: start -->
<br>

## About *rankinma*

Package *rankinma* supports users to easily obtain and visualize various metrics of treatment ranking from network meta-analysis without major concerns regarding heterogeneity, intransitivity, and incoherence. This package not only accepts manual-prepared data set of treatment ranking metrics from users, but also can help users to gather various treatment ranking metrics in network meta-analysis. 

<!-- content: end -->
<!-- features: start -->
<br>


## Feature

*rankinma* allows users to visualize various treatment ranking metrics in network meta-analysis based either common-effect model or random-effects model no matter using frequentist or Bayesian approach. The current version includes three common metrics of treatment ranking.

-   **Probabilities:** probabilities of every available treatment on each possible rank.
-   **SUCRA:** the value of surface under the cumulative ranking curve using Bayesian approach.
-   **P-score:** the value of SUCRA using frequentist approach.

<br>

Briefly, *rankinma* can be used for visualization of both detailed metrics of probabilities and global metrics (i.e. SUCRA and P-score). Besides, *rankinma* provides users multiple types of plots to illustrate aforementioned treatment ranking metrics, and current version consists of five types of plots with six sub-types.

-   **Beading plot:** a novel graphics for displaying global metrics of treatment ranking (i.e. SUCRA and P-score) based on numeric line plot.
-   **Bar chart:** a classic graphics for most metrics of treatment ranking (i.e. probabilities, SUCRA, and P-score), and *rankinma* supports two sub-type of bar chart in terms of side-by-side bar chart and cumulative bar chart.
-   **Line chart:** a classic graphics for most metrics of treatment ranking (i.e. probabilities, SUCRA, and P-score), and *rankinma* supports two sub-type of line chart in terms of simple line chart (a line on a chart) and composite line chart (multiple lines on a chart).
-   **Heat plot:** a new graphics for showing global metrics of treatment ranking (i.e. SUCRA and P-score) for each outcome, and *rankinma* supports to gather all heat plots of outcomes with interests on a plot.
-   **Spie chart:** a new graphics proposed in 2020 for displaying multiple global metrics of treatment ranking (i.e. SUCRA and P-score) from outcomes with interests by each treatment, and *rankinma* supports to place all spie charts on a plot.

<!-- features: end -->
<!-- installation: start -->
<br>


## Installation

Package *rankinma* has been formally released on the [CRAN](https://cran.r-project.org/web/packages/rankinma/index.html), and users can install *rankinma* in R using following syntax:

```{r}
install.packages("rankinma")
```

<br>

If users have tabulated global metrics of treatment ranking in a csv file, they can illustrate beading plot without installation of package *rankinma* by using online platform [**RankiNMA** (click here)](https://rankinma.shinyapps.io/rankinma/).
    
<!-- installation: end -->
<!-- examples: start -->
<br>


## Examples

### Example 1

To depict global metrics of treatment ranking using bar chart:

<p align = "center" width = "100%">
<img src = "https://github.com/EnochKang/rankinma/blob/main/vignettes/Figure%201A.png?raw=true">
</p>
<br>

### Example 2

To depict probabilities of treatments on each possible rank using multi-line chart:

<p align = "center" width = "100%">
<img src = "https://github.com/EnochKang/rankinma/blob/main/vignettes/Figure%202A.png?raw=true">
</p>
<br>

### Example 3

To depict probabilities of treatments on each possible rank using stacked bar chart:

<p align = "center" width = "100%">
<img src = "https://github.com/EnochKang/rankinma/blob/main/vignettes/Figure%202B.png?raw=true">
</p>
<br>

### Example 4 

To depict global metrics of treatment ranking using beading plot:

<p align = "center" width = "100%">
<img src = "https://github.com/EnochKang/rankinma/blob/main/vignettes/Figure%203.png?raw=true">
</p>
<br>
    
<!-- examples: end -->
<!-- coding conventions: start -->
<br>


## Coding conventions

This package is mainly written according to [Google's R style](https://web.stanford.edu/class/cs109l/unrestricted/resources/google-style.html). For readers, details of naming rules are listed as follows:

1. **.R file** is named using lower case with underscore "_" between words (*e.g. get_metrics.R*). 

2. **function** is named using verb or verb with noun, and the first character of each word is written in capital letter (e.g. `GetMetrics()`).

3. **object** is named using noun with the first word in lower case, but the first character of rest words is written using capital letter (e.g. `dataBeads`).

4. **variable** is named using noun written in lower case. Words of variable name are separated by "." if a variable name consists of more than two words (e.g. `dataRankinma$metrics.name`).

<br>


Version number consists of three integers with a period between them (eg. version 1.0.0).

1. Updating the first integer refers to a modification with new methodological impact. 

2. Changing the second integer refers to an update with a new function without new methodological impact. 

3. Updating the third integer refers to a modification in a function.

<!-- coding conventions: end -->
<!-- license: start -->
<br>


## License

This package is licensed under the [GPL-3 License](https://cran.r-project.org/web/licenses/GPL-3).

<!-- license: end -->
<!-- to to list: start -->
<br>


## To do list

Task force will keep update package *rankinma* for following issues:

-   Scatter plot
-   User-designed color gradient for heat plot.

<!-- to to list: end -->


NEWS.md
# Change of package `rankinma`


## rankinma 0.1.9
### Date: August 31, 2023

**Modify document:** improve *rankinma* package README file.
\

### Date: August 20, 2023

**Modify document:** improve *rankinma* package manual by detailing returned values from function `SetMetrics()`.
\

### Date: August 08, 2023

**Modify document:** improve *rankinma* package manual by displaying description and details.
\

**Modify function:** improve function `GetMetrics()` in order to get SUCRA consistently and reproducibly.
\

### Date: July 31, 2023

**Modify function:** improve function `PlotBeads()` with newly added parameter `szPnt`, `szFntTtl`, `szFntTtlX`, `szFntX`, `szFntY`, and `szFntLgnd` for allowing user to change size of beads, font size of main title for beading plot, font size of title on X-axis, font size of numeric scale on X-axis, font size of outcome name(s), and legend font size. Minor modification was also done for sorting legend by alphabet.
\

**Modify function:** improve function `PlotHeat()` with a newly added parameter `szFntY` for allowing user to change font size of outcome name(s).
\

### Date: July 30, 2023

**Modify function:** improve function `PlotLine()` and `PlotBar()` to illustrate line chart or bar chart according to descending order of global metrics of treatment ranking on single outcome.
\

**Modify function:** improve function `PlotLine()`, `PlotBar()`, and `PlotHeat()` with a newly added parameter `rotateX` in order to rotate labels on x-axis.
\



## rankinma 0.1.8
### Date: July 03, 2023

**Update to CRAN**  
\

**Modify function:** remove routinely reported processing information from all functions in the *rankinma* package. The information only appear when users use function(s) inappropriately.
\

**Modify function:** improve function `PlotLine()` in order to produce cumulative line charts, and merged line chart.
\

**Modify function:** improve function `PlotLine()` in order to produce cumulative line charts, and merged line chart.
\

### Date: June 23, 2023

**Modify function:** improve data setting in function `SetMetrics()` for scatter plot.
\

**Modify function:** improve data setting in both `GetMetrics()` and `SetMetrics()` functions for cumulative probability of rank.
\


## rankinma 0.1.7
### Date: April 23, 2023

**Update to CRAN**  
\

### Date: April 22, 2023

**Modify function:** add transparency in both functions of `SetMetrics()` and `PlotBeads()`.
\

**Modify function:** modify function `ShowColor()` based on the updated function `SetMetrics()` on April 20, 2023.
\

**Modify function:** improve function `PlotBar()`, `PlotLine()`, and `PlotSpie()` to show user-defined colors appropriately.
\

**Modify function:** improve function `PlotHeat()` to show heat plot with order of metrics or treatment appropriately.
\


## rankinma 0.1.6
### Date: April 20, 2023

**Add function:** `PlotSpie()` is added to the *rankinma* in order to illustrate Spie plot of global metrics of treatment ranking (i.e. SUCRA and P-score) using object of class *rankinma* set by function `SetMetrics()`.
\

**Modify function:** improve data setting and color in both functions of `SetMetrics()` and `PlotBeads()`.
\

**Update to CRAN**  
\


## rankinma 0.1.5
### Date: April 4, 2023

**Modify function:** keep user's options of par (*'par(no.readonly)'*) after illustrating plots with *rankinma* by using function `on.exit()`.
\

**Re-submit to CRAN**  
\

### Date: April 1, 2023

**Add function:** `PlotHeat()` is added to the *rankinma* in order to illustrate heat plot of global metrics of treatment ranking (i.e. SUCRA and P-score) using object of class *rankinma* set by function `SetMetrics()`.
\


## rankinma 0.1.4
### Date: March 27, 2023

**Modify function:** remove global assignment (*'<<-'*) from package.
\

**Submit to CRAN**  
\

### Date: February 19, 2023

**Push to github**  
\

### Date: February 10, 2023

**Modify function:** all functions were modified due to warnings regarding *... invisible binding for '<<-' assignment ...* .
\


## rankinma 0.1.3
### Date: February 06, 2023

**Add function:** `PlotLine()` is added to the *rankinma* in order to illustrate line chart of global metrics of treatment ranking (i.e. SUCRA and P-score) using object of class *rankinma* set by function `SetMetrics()`.
\


## rankinma 0.1.2
### Date: January 31, 2023

**Add function:** `PlotBar()` is added to the *rankinma* in order to illustrate bar chart of global metrics of treatment ranking (i.e. SUCRA and P-score) using object of class *rankinma* set by function `SetMetrics()`.
\


## rankinma 0.1.1 
### Date: January 30, 2023

**Add function:** `GetMetrics()` is added to the *rankinma* in order to get metrics of treatment ranking from object of class *netmeta*.
\


## rankinma 0.1.0
### Date: January 29, 2023

Build rankinma with first two functions including `SetMetrics()` and `PlotBeads()`.

> 1. `SetMetrics()` is to set imported data in *rankinma* format.
>
> 2. `PlotBeads()` is to illustrate beading plot using object of class *rankinma* set by function `SetMetrics()`.
>


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(rankinma)

test_check("rankinma")


R/plot_beads.R
#' @title Illustrate beading plot
#'
#' @author Chiehfeng Chen & Enoch Kang
#'
#' @description
#' **PlotBeads()** is a function for illustrating beading plot.
#'
#' @param data DATA of metrics for treatment ranking.
#' @param color LIST of colors for treatments in a network meta-analysis.
#'
#' @return
#' **PlotBeads()** returns a beading plot.
#'
#' @seealso \code{\link{GetMetrics}} \code{\link{SetMetrics}}
#'
#' @examples
#' ## Not run:
#' library(netmeta)
#' data(Senn2013)
#' nma <- netmeta(TE, seTE, treat1, treat2,
#'   studlab, data = Senn2013, sm = "SMD")
#'
#' # Get SUCRA
#' nma.1 <- GetMetrics(nma, outcome = "HbA1c.random", prefer = "small", metrics = "SUCRA",
#'   model = "random", simt = 1000)
#' nma.2 <- GetMetrics(nma, outcome = "HbA1c.common", prefer = "small", metrics = "SUCRA",
#'   model = "common", simt = 1000)
#'
#' # Combine metrics of multiple outcomes
#' dataMetrics <- rbind(nma.1, nma.2)
#'
#' # Set data for rankinma
#' dataRankinma <- SetMetrics(dataMetrics, tx = tx, outcome = outcome,
#'   metrics = SUCRA, metrics.name = "SUCRA")
#'
#' # Illustrate beading plot
#' PlotBeads(data = dataRankinma)
#' ## End(Not run)
#'
#' @export PlotBeads

PlotBeads <- function(data,
                      color = NULL) {

  dataBeads <- data$data

  lgcInher  <- !inherits(data, "rankinma")
  lgcMtrcs  <- ifelse(data$metrics.name == "Probabilities",
                      TRUE,
                      FALSE)
  
  lgcColor  <- ifelse(length(which(ls()%in%ls(pattern = "color"))) > 0, FALSE,
                      ifelse(length(color) != data$n.tx,
                             TRUE, FALSE))

  if (lgcInher) {
    cat(paste(" Inherit -------------------------------------------------- X\n",
              ' REQUIRE: Argument "data" must be an object of class \"rankinma\".'),
        fill = TRUE, sep = "")
  } else {
    cat(paste(" Inherit -------------------------------------------------- V"),
        fill = TRUE, sep = "")
  }

  if (lgcMtrcs) {
    cat(paste(" Metrics -------------------------------------------------- X\n",
              ' REQUIRE: Metrics should not be "Probabilities."'),
        fill = TRUE, sep = "")
  } else {
    cat(paste(" Metrics -------------------------------------------------- V"),
        fill = TRUE, sep = "")
  }

  if (lgcColor) {
    cat(paste(" Color ---------------------------------------------------- X\n",
              ' REQUIRE: Argument "color" must list colors for **EACH TREATMENT**.'),
        fill = TRUE, sep = "")
  } else {
    cat(paste(" Color ---------------------------------------------------- V"),
        fill = TRUE, sep = "")
  }

  if (lgcInher | lgcMtrcs | lgcColor)
    stop("Correct above mentioned problem(s).")

  dataBeads$importance  <- dataBeads$outcomes
  dataBeads$shape       <- 16

  dataBeads$seq.outcome <- max(dataBeads$outcomes) + 1 - dataBeads$outcomes
  dataBeads$seq.tx      <- max(dataBeads$txs) + 1 - dataBeads$txs

  dataBeads$seq.axis.y  <- dataBeads$seq.outcome
  dataBeads             <- dataBeads[order(-dataBeads$seq.axis.y), ]

  txs      <- unique(dataBeads$tx)
  outcomes <- unique(dataBeads$outcome)

  dataBeadsPlot <- dataBeads
  
  txColor <- data$color.txs
  
  if (!is.null(color)) {
    if (length(which(ls()%in%ls(pattern = "color"))) > 0) {
      txColor$color.tx <- color
      for (color.i in c(1:nrow(dataBeadsPlot))) {
        dataBeadsPlot[color.i, "color.tx"] <- txColor[which(dataBeadsPlot[color.i, "tx"] == txColor$lsTx), "color.tx"]
      }
    }
  }

  setPar <- par(no.readonly = TRUE)
  on.exit(par(setPar))

  par(mar = c(5, 5, 3, 5), xpd = TRUE)

  plot(dataBeadsPlot$metrics,
       dataBeadsPlot$seq.axis.y - 0.5, frame.plot = FALSE,
       xlim = c(-0.3, 1.3),
       ylim = c(0, ceiling(max(dataBeadsPlot$importance, na.rm = TRUE))),
       xlab = "", xaxt = "n", yaxt = "n",ylab="",
       pch = 0, cex = 0)
  axis(side = 1, at = c(0, 0.2, 0.4, 0.6, 0.8,1), cex.axis = 0.8)
  axis(side = 3, at = c(0.5), line = -2, tick = FALSE,
       labels = paste("Beading plot of ", data$metrics.name,
                      sep = ""),
       font = 2, cex.axis = 1.2)
  axis(side = 1, at = c(0.5), line = 2, tick = FALSE,
       labels = paste("Worse <---   ", data$metrics.name,
                      "   ---> Better", sep = ""),
       font = 1, cex.axis = 1)
  text(rep(-0.3, data$n.outcome),
       unique(dataBeadsPlot$seq.outcome) - 0.5,
       c(unique(dataBeadsPlot$outcome)), pos = 4, cex = 1)
  segments(0, c(dataBeadsPlot$outcomes) - 0.5,
           1, c(dataBeadsPlot$outcomes) - 0.5, col = "gray",
           lty = c(rep(1, data$n.outcome)))

  points(dataBeadsPlot$metrics,
         dataBeadsPlot$seq.axis.y - 0.5,
         pch = dataBeadsPlot$shape,
         col = c(dataBeadsPlot$color.tx),
         bg = c(dataBeadsPlot$color.tx), cex = 2)
  points(c(rep(1.1, data$n.tx)),
         (data$n.outcome - 0.5) / data$n.tx * c(1:data$n.tx),
         pch = dataBeadsPlot$shape,
         col = c(unique(dataBeadsPlot$color.tx)),
         bg = c(unique(dataBeadsPlot$color.tx)))
  
  text(c(rep(1.15, data$n.tx)),
       (data$n.outcome - 0.5) / data$n.tx * c(1:data$n.tx),
       unique(dataBeadsPlot$tx), pos = 4)
}


R/rankinma-package.R
#' rankinma: Ranking in Network Meta-Analysis
#'
#' @description
#' \emph{rankinma} is an R package that supports users to easily obtain and
#' visualize various metrics of treatment ranking from network meta-analysis
#' no matter using either frequentist or Bayesian approach. Development of package
#' \emph{rankinma} is based on \bold{R version 4.2.2 (2022-10-31 ucrt)}.
#' Extra imported packages are as follows:
#'
#' \itemize{
#'  \item \href{https://cran.r-project.org/src/contrib/Archive/mvtnorm/mvtnorm_1.1-3.tar.gz}{\emph{mvtnorm} (version 1.1-3)}
#'  \item \href{https://cran.r-project.org/src/contrib/Archive/netmeta/netmeta_2.6-0.tar.gz}{\emph{netmeta} (version 2.6-0)}
#' }
#'
#'
#' @details
#' Current version consists of seven functions, including two functions for
#' data preparation (function \code{\link{GetMetrics}} and \code{\link{SetMetrics}})
#' and five functions for visualization of treatment ranking metrics (i.e.
#' \code{\link{PlotBeads}}, \code{\link{PlotLine}}, \code{\link{PlotBar}},
#' \code{\link{PlotHeat}}, and \code{\link{PlotSpie}}). Probabilities of treatments
#' on each possible rank can be visualized using \code{\link{PlotLine}} and \code{\link{PlotBar}}.
#' Due to concise information, \code{\link{PlotBeads}} is recommended to be used for
#' global metrics of treatment ranking, such as \bold{P-score} and \bold{SUCRA}. The other four
#' visualization functions can also generate graphics of the global metrics.
#'
#' @name rankinma-package
#'
#' @docType package
#'
#' @keywords package
#'
#'
## usethis namespace: start
#' @importFrom graphics axis
#' @importFrom graphics barplot
#' @importFrom graphics lines
#' @importFrom graphics mtext
#' @importFrom graphics par
#' @importFrom graphics pie
#' @importFrom graphics points
#' @importFrom graphics rect
#' @importFrom graphics segments
#' @importFrom graphics text
#' @importFrom graphics title
#' @importFrom grDevices col2rgb
#' @importFrom grDevices rainbow
#' @importFrom grDevices recordPlot
#' @importFrom grDevices rgb
#' @importFrom stats aggregate
#' @importFrom utils capture.output
## usethis namespace: end
NULL


R/set_metrics.R
#' @title Setup data of treatment ranking metrics for rankinma
#'
#' @author Enoch Kang
#'
#' @description
#' **SetMetrics()** is a function for checking and preparing data set of metrics
#' for further ploting in *rankinma*.
#'
#' @param data DATAFRAME of treatment, metrics, and name of outcomes.
#' @param outcome VARIABLE string data for of outcome(s).
#' @param tx VARIABLE with string data for treatments.
#' @param metrics VARIABLE with numeric data for global metrics, but it should
#'        be "NULL" when using "Probabilities" as metrics.
#' @param metrics.name STRING for metrics of treatment ranking in terms of
#'        "SUCRA" and "P-score" for the value of surface under the cumulative
#'        ranking curve or P-score.
#'
#' @return
#' **SetMetrics()** returns a confirmed data.frame of treatment, metrics of
#' treatment ranking, and outcome name.
#'
#' @seealso \code{\link{GetMetrics}}
#'
#' @examples
#' ## Not run:
#' library(netmeta)
#' data(Senn2013)
#' nma <- netmeta(TE, seTE, treat1, treat2,
#'   studlab, data = Senn2013, sm = "SMD")
#'
#' # Get SUCRA
#' nma.1 <- GetMetrics(nma, outcome = "HbA1c.random", prefer = "small", metrics = "SUCRA",
#'   model = "random", simt = 1000)
#' nma.2 <- GetMetrics(nma, outcome = "HbA1c.common", prefer = "small", metrics = "SUCRA",
#'   model = "common", simt = 1000)
#'
#' # Combine metrics of multiple outcomes
#' dataMetrics <- rbind(nma.1, nma.2)
#'
#' # Set data for rankinma
#' dataRankinma <- SetMetrics(dataMetrics, tx = tx, outcome = outcome,
#'   metrics = SUCRA, metrics.name = "SUCRA")
#' ## End(Not run)
#'
#' @export SetMetrics

SetMetrics <- function(data,
                       outcome = NULL,
                       tx = NULL,
                       metrics = NULL,
                       metrics.name = NULL) {
  
  data <- data.frame(data)
  lsMetrics <- c("Probabilities", "P-best", "SUCRA", "P-score")

  namTx      <- deparse(substitute(tx))
  namMetrics <- deparse(substitute(metrics))
  namOutcome <- deparse(substitute(outcome))

  lgcTx      <- !(namTx %in% colnames(data))
  lgcOutcome <- !(namOutcome %in% colnames(data))
  lgcMetrics <- ifelse(metrics.name == "Probabilities",
                       FALSE,
                       !(namMetrics %in% colnames(data)))
  lgcDupl    <- ifelse(sum(table(data$tx, data$outcome)) == sum(table(data$tx, data$outcome) == 1),
                       FALSE, TRUE)

  cat("Check variables:\n")

  if (!(namTx %in% colnames(data))) {
    cat(paste(" Treatment -------------------------------------------------- X\n",
              ' REQUIRE: Argument "tx" must be specified.\n'),
        fill = TRUE, sep = "")
  } else {
    colnames(data)[which(colnames(data) == namTx)] <- "tx"
    cat(paste(" Treatment -------------------------------------------------- V"),
        fill = TRUE, sep = "")
  }

  if (!(namOutcome %in% colnames(data))) {
    cat(paste(" Outcome   -------------------------------------------------- X\n",
              ' REQUIRE: Argument "outcome" must be specified.'),
        fill = TRUE, sep = "")
  } else {
    colnames(data)[which(colnames(data) == namOutcome)] <- "outcome"
  }

  if (isFALSE(sum(table(data$tx, data$outcome)) == sum(table(data$tx, data$outcome) == 1))) {
    cat(paste(" Outcome   -------------------------------------------------- X\n",
              ' REQUIRE: Duplicates of treatment in a same outcome.'),
        fill = TRUE, sep = "")
  } else {
    cat(paste(" Outcome   -------------------------------------------------- V"),
        fill = TRUE, sep = "")
  }

  if (is.null(metrics.name)) {
    cat(paste(" Metrics name  ---------------------------------------------- X\n",
              ' REQUIRE: Argument "metrics.name" must be "Probabilities,"
              "SUCRA," or "P-score."'),
        fill = TRUE, sep = "")
  } else if (!(metrics.name %in% lsMetrics)) {
    cat(paste(" Metrics name  ---------------------------------------------- X\n",
              ' REQUIRE: Argument "metrics.name" must be "Probabilities,"
              "SUCRA," or "P-score."'),
        fill = TRUE, sep = "")
  } else {
    cat(paste(" Metrics name  ---------------------------------------------- V"),
        fill = TRUE, sep = "")
  }

  if (metrics.name != "Probabilities") {
    if (!(namMetrics %in% colnames(data))) {
      cat(paste(" Metrics   -------------------------------------------------- X\n",
                ' REQUIRE: Argument "metrics" must be specified for SUCRA or P-score.'),
          fill = TRUE, sep = "")
    } else {
      colnames(data)[which(colnames(data) == namMetrics)] <- "metrics"
      if (!is.numeric(data$metrics)) {
        cat(paste(" Metrics   -------------------------------------------------- X\n",
                  ' REQUIRE: Argument "metrics" must be numeric data for SUCRA
                  or P-score.'),
            fill = TRUE, sep = "")
      } else {
        cat(paste(" Metrics   -------------------------------------------------- V"),
            fill = TRUE, sep = "")
      }
    }
  } else if (isFALSE(is.null(namMetrics))) {
    cat(paste(" Metrics   -------------------------------------------------- !\n",
              ' INFORM: Argument "metrics" is not mandatory for "Probabilities,
              and *rankinma* ignores your argument "', namMetrics,  '".',
              sep = ""),
        fill = TRUE, sep = "")
  }

  if (metrics.name == "Probabilities") {
    if (lgcTx |
        lgcOutcome |
        !(metrics.name %in% lsMetrics) |
        lgcMetrics |
        lgcDupl) {
      lgcOverall <- TRUE
    } else {
      lgcOverall <- FALSE
    }
  } else if (lgcTx |
             lgcOutcome |
             !(metrics.name %in% lsMetrics) |
             lgcMetrics |
             !is.numeric(data$metrics) |
             lgcDupl) {
    lgcOverall <- TRUE
  } else {
    lgcOverall <- FALSE
  }

  if (lgcOverall)
    stop("Correct above mentioned problem(s).")

  txs      <- unique(data$tx)
  outcomes <- unique(data$outcome)

  for (i in c(1:nrow(data))) {
    data[i, "txs"] <- which(as.character(txs) == as.character(data[i, "tx"]))
  }
  for (i in seq(1, nrow(data))) {
    data[i, "outcomes"] <- which(as.character(outcomes) == as.character(data[i, "outcome"]))
  }

  data$importance <- max(data$outcomes) + 1 - data$outcomes

  data$color.r <- NA
  data$color.g <- NA
  data$color.b <- NA

  txColor <- data.frame(lsTx   = unique(data$tx),
                        seqTx  = c(1:length(unique(data$tx))),
                        color.r = NA,
                        color.g = NA,
                        color.b = NA)
  txColor$color.base <- (max(txColor$seqTx) - txColor$seqTx^0.99) / max(txColor$seqTx)
  for (color.i in seq(from = 1, to = nrow(txColor), by = 3)) {
    txColor[color.i, "color.r"] <- txColor$color.base[color.i]
    txColor[color.i, "color.g"] <- txColor$color.base[max(txColor$seqTx)+1-color.i]
    txColor[color.i, "color.b"] <- txColor$color.base[max(txColor$seqTx)+1-color.i]
  }
  for (color.i in seq(from = 2, to = nrow(txColor), by = 3)) {
    txColor[color.i, "color.r"] <- txColor$color.base[max(txColor$seqTx)+1-color.i]
    txColor[color.i, "color.g"] <- txColor$color.base[color.i]
    txColor[color.i, "color.b"] <- txColor$color.base[max(txColor$seqTx)+1-color.i]
  }
  for (color.i in seq(from = 3, to = nrow(txColor), by = 3)) {
    txColor[color.i, "color.r"] <- txColor$color.base[max(txColor$seqTx)+1-color.i]
    txColor[color.i, "color.g"] <- txColor$color.base[max(txColor$seqTx)+1-color.i]
    txColor[color.i, "color.b"] <- txColor$color.base[color.i]
  }
  
  txColor$color.r1 <- txColor$color.r
  txColor$color.g1 <- txColor$color.g
  txColor$color.b1 <- txColor$color.b

  txColor$color.r2 <- 1-txColor$color.r^2*1.3
  txColor$color.g2 <- 1-txColor$color.g^2*1.3
  txColor$color.b2 <- 1-txColor$color.b^2*1.3

  txColor$color.r <- 1-(1-txColor$color.r)^1.2
  txColor$color.g <- txColor$color.g^0.3
  txColor$color.b <- 1-(1-txColor$color.b)^2
  
  for (color.i in c(1:nrow(data))) {
    data[color.i, "color.r"] <- txColor[which(data[color.i, "tx"] == txColor$lsTx), "color.r"]
    data[color.i, "color.g"] <- txColor[which(data[color.i, "tx"] == txColor$lsTx), "color.g"]
    data[color.i, "color.b"] <- txColor[which(data[color.i, "tx"] == txColor$lsTx), "color.b"]
  }
  
  #plot(7, 0.5, xlim = c(1, 10), ylim = c(0, 1),
  #     frame = F, col = rgb(1, 1, 1, 1))
  #points(txColor$seqTx, txColor$color.base, pch = 16,
  #       col = rgb(1-(1-txColor$color.r)^1.2,
  #                 txColor$color.g^0.3, 
  #                 1-(1-txColor$color.b)^2,
  #                 0.9))
  #points(txColor$seqTx, txColor$color.base, pch = 16,
  #       col = rgb(1-1/(i^0.7),
  #                 (max(i)-i^0.99)/max(i), 
  #                 1/(i^0.3),
  #                 0.9))
  
  sink(tempfile())
  data$color.tx <- print(rgb(data$color.r,
                             data$color.g,
                             data$color.b, 0.8))
  sink()


  if (metrics.name == "Probabilities") {
    dataSet <- as.data.frame(cbind(
      data[, c("outcome", "outcomes", "tx", "txs", "rank", "color.tx")],
      data[, c(3:(length(unique(data$tx))+2))]))
  } else {

    dataSet <- data[, c("outcome", "outcomes", "importance", "tx", "txs", "color.tx", "metrics")]

    for (outcome.i in c(1:max(dataSet$outcomes))) {
      dataTempA <- dataSet[dataSet$outcomes == outcome.i, ]

      if (length(which(as.data.frame(table(dataTempA$metrics))$Freq > 1)) > 0) {
        for (tx.i in c(1:(nrow(dataTempA) - 1))) {
          if (dataTempA$importance[tx.i] %in% dataTempA$importance[c((tx.i + 1):nrow(dataTempA))]) {
            importanceAdj <- sum(dataTempA$metrics[c((tx.i + 1):nrow(dataTempA))] == dataTempA$metrics[tx.i]) / 20
            dataTempA$importance[tx.i] <- importanceAdj + dataTempA$importance[tx.i]
          }
        }
      }

      if (outcome.i == 1) {
        dataTempB <- dataTempA
      } else {
        dataTempB <- rbind(dataTempA, dataTempB)
      }
    }
    dataSet <- dataTempB

  }

  dataList        <- list(metrics.name = metrics.name,
                          ls.outcome   = unique(data$outcome),
                          ls.tx        = unique(data$tx),
                          n.outcome    = length(unique(data$outcome)),
                          n.tx         = length(unique(data$tx)))
  class(dataList)    <- "rankinma"
  dataList$data      <- dataSet
  dataList$data.sets <- split(dataSet, dataSet$outcome)
  dataList$color.txs <- txColor
  dataRankinma       <- dataList


  cat(paste("\n"), fill = TRUE, sep = "")
  cat(paste("Summary of metrics:\n",
            "Metrics: ", dataList$metrics.name, "\n",
            "Outcomes:   ", dataList$n.outcome, "\n",
            "Treatments: ", dataList$n.tx, "\n"),
      fill = TRUE, sep = "")

  cat(paste("List of outcomes:"),
      fill = TRUE, sep = "")
  cat(paste(" ", c(1:dataList$n.outcome), dataList$ls.outcome, sep = " "),
      fill = TRUE, sep = "\n")

  cat(paste("List of treatments:"),
      fill = TRUE, sep = "")
  cat(paste(" ", c(1:dataList$n.tx), dataList$ls.tx, sep = " "),
      fill = TRUE, sep = "\n")

  dataRankinma <- dataList

}


#' @title Display color for each treatment
#'
#' @author Enoch Kang
#'
#' @description
#' ShowColor is a function for showing colors of every treatment on plot of
#' treatment rank metrics in *rankinma*.
#'
#' @param data DATA of *rankinma* class.
#'
#' @return
#' **ShowColor()** show a plot of color for each treatment.
#'
#' @export ShowColor

ShowColor <- function(data) {

  if (!inherits(data, "rankinma"))
    stop('Argument "data" must be an object of class \"rankinma\".')

  plot(c(0, 1.5),
       c(0, 1),
       frame.plot = FALSE,
       xlab = "", ylab="", xaxt = "n", yaxt = "n",
       col = "white")
  segments(rep(0.5, max(data$txs)),
           (0.95) / max(data$txs) * unique(data$txs),
           rep(0.7, max(data$txs)),
           (0.95) / max(data$txs) * unique(data$txs),
           lwd = 6,
           col = data$color.tx)
  text(rep(0.8, max(data$txs)),
       (0.95) / max(data$txs) * unique(data$txs),
       data$tx, pos = 4)
}
